onnxscript/version_converter/_version_converter.py

```diff
@@ -0,0 +1,337 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+"""Convert the model to the specified ONNX opset version."""
+
+from __future__ import annotations
+
+import dataclasses
+import functools
+import logging
+from typing import Callable, Sequence, Union
+
+import onnxscript.ir.convenience as ir_convenience
+import onnxscript.rewriter.pattern as orp
+from onnxscript import ir
+
+logger = logging.getLogger(__name__)
+
+
+CURRENT_MAX_ONNX_OPSET = 23
+
+
+class VersionConverterError(RuntimeError):
+    """Raised when an node's version cannot be upgraded/downgraded successfully."""
+
+
+@dataclasses.dataclass
+class Replacement:
+    """A replacement for a node in the graph."""
+
+    new_outputs: Sequence[ir.Value]
+    new_nodes: Sequence[ir.Node]
+
+
+# A version-adapter function takes a node, a RewriterContext and returns
+# a Replacement for the node or None (if no replacement is needed).
+
+ReturnValue = Union[Sequence[ir.Value], ir.Value, None]
+AdapterFunction = Callable[[ir.Node, orp.RewriterContext], ReturnValue]
+
+
+@dataclasses.dataclass
+class VersionAdapter:
+    """A class that represents a version checker for a particular op.
+
+    It is applicable for a specific version upgrade (orignal_version -> original_version + 1)
+    or downgrade (orignal_version -> original_version - 1)of the op.
+    """
+
+    node_version: int
+    up_conversion: bool
+    function: AdapterFunction
+
+
+class AdapterRegistry:
+    """A class that maintains a registry of adapters for ops."""
+
+    def __init__(self):
+        self.op_adapters: dict[tuple[str, str, int, bool], VersionAdapter] = {}
+
+    def lookup_adapters(
+        self,
+        domain: str,
+        opname: str,
+        original_version: int,
+        up_conversion: bool = True,
+    ) -> AdapterFunction | None:
+        adapter = self.op_adapters.get((domain, opname, original_version, up_conversion))
+        if adapter is not None:
+            return adapter.function
```

Comment:
This is the only place where the adapter is used, and we just return adapter.function. So, do we really need the `VersionAdapter` class? The map can just store the "function" directly, which will also get rid the extra 3 lines here.